### PR TITLE
integrate RSocketEvents

### PR DESCRIPTION
**Pull request by IdoJS**
Integrate RSocketEventsClient & RSocketEventsServer to allowe event base communication between proxy and remoteMicroservice.

### CODE REVIEW

File: packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts [hunk 1]
```diff
@@ -26,7 +28,71 @@ export const remoteCall = ({
...
+        const serializeData = JSON.stringify(message);
+        const socketConnect = socket[asyncModel]({
+          data: serializeData,
+          metadata: '',
+        });
+
+        switch (asyncModel) {
+          case ASYNC_MODEL_TYPES.REQUEST_RESPONSE:
+            socketConnect.then((response: RsocketEventsPayload) => {
+              const { data } = response && JSON.parse(response.data);
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts]: the callback for both async models is the same - I suggest to extract it to a common helper

File: packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts [hunk 2]
```diff
@@ -26,7 +28,71 @@ export const remoteCall = ({
...
+  if (!openConnections[address]) {
+    const client = CreateClient({ address });
+    connection = client.connect();
+    openConnections[address] = connection;
+  } else {
+    connection = openConnections[address];
+  }
+
+  return new Observable((observer) => {
+    try {
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts]: why are you using try/catch for promise handling? seems like catch for the promise in this case is appropriate

File: packages/scalecube-microservice/src/ServiceCall/ServiceCall.ts [hunk 3]
```diff
@@ -4,8 +4,11 @@ import { throwErrorFromServiceCall } from '../helpers/utils';
 import { MESSAGE_NOT_PROVIDED, ASYNC_MODEL_TYPES } from '../helpers/constants';
 import { localCall } from './LocalCall';
 import { remoteCall } from './RemoteCall';
+import { take } from 'rxjs/operators';
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/ServiceCall/ServiceCall.ts]: please put absolute imports before the relative

File: packages/scalecube-microservice/src/TransportProviders/MicroserviceServer.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  metadata,
+  serviceCall,
+}: {
+  data: string;
+  metadata: string;
+  serviceCall: ServiceCall;
+}) => {
+  return new Single((subscriber: any) => {
+    subscriber.onSubscribe();
+    const message = JSON.parse(data);
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/TransportProviders/MicroserviceServer.ts]: what if JSON.parse will throw an error?

File: packages/scalecube-microservice/src/TransportProviders/MicroserviceServer.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  metadata,
+  serviceCall,
+}: {
+  data: string;
+  metadata: string;
+  serviceCall: ServiceCall;
+}) => {
+  return new Flowable((subscriber: any) => {
+    subscriber.onSubscribe();
+    const message = JSON.parse(data);
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/TransportProviders/MicroserviceServer.ts]: the same here

File: packages/scalecube-microservice/src/helpers/types.ts [hunk 6]
```diff
@@ -109,3 +110,8 @@ export interface MicroserviceContext {
   serviceRegistry: ServiceRegistry;
   methodRegistry: MethodRegistry;
 }
+
+export interface RsocketEventsPayload {
+  data: any;
+  metadata: any;
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/helpers/types.ts]: can't we specify the metadata better?

File: packages/scalecube-microservice/tests/integration/Errors/errors-create-serviceCall.spec.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+describe('describe', () => {
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/tests/integration/Errors/errors-create-serviceCall.spec.ts]: Is it a placeholder for now?

File: packages/scalecube-microservice/tests/integration/properties-validation.spec.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
+describe('describe', () => {
+  test('test', () => {
+    expect(true).toBe(true);
+  });
+});
+// /**
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/tests/integration/properties-validation.spec.ts]: What will we do with this test?

Comment (IdoJS) [reply packages/scalecube-microservice/tests/integration/properties-validation.spec.ts]: placeholder for validations

File: packages/scalecube-microservice/src/helpers/types.ts (hunk 6 above)
Comment (IdoJS) [reply packages/scalecube-microservice/src/helpers/types.ts]: atm we aren't using the metadata value

File: packages/scalecube-microservice/package.json [hunk 9]
```diff
@@ -18,7 +18,7 @@
     "build": "rimraf .cache && npm-run-all build:commonJS build:es build:umd build:umd-min",
     "build:commonJS": "rimraf lib && tsc --module commonjs --target es5 --outDir lib",
     "build:es": "rimraf es && tsc --module esnext --target es5 --outDir es",
-    "build:umd": "rimraf dist && parcel build src/index.ts --no-minify --experimental-scope-hoisting --out-dir dist dist --global sc",
+    "build:umd": "rimraf dist && parcel build src/index.ts --no-minify  --out-dir dist dist --global sc",
```
[the remark this answers is not in the document]
Comment (IdoJS) [reply packages/scalecube-microservice/package.json]: rsocket doesn't support --experimental-scope-hoisting

File: packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts [hunk 10]
```diff
@@ -42,59 +44,58 @@ const remoteResponse = ({
...
-
-        switch (asyncModel) {
-          case ASYNC_MODEL_TYPES.REQUEST_RESPONSE:
-            socketConnect.then((response: RsocketEventsPayload) => {
-              const { data } = response && JSON.parse(response.data);
-              observer.next(data);
-            });
-            // TODO Add catch
-            break;
+    let connection: any;
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts]: connection is a Promise type

File: packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts [hunk 11]
```diff
@@ -42,59 +44,58 @@ const remoteResponse = ({
...
+    }
 
-          case ASYNC_MODEL_TYPES.REQUEST_STREAM:
-            socketConnect.subscribe((response: RsocketEventsPayload) => {
-              const { data } = response && JSON.parse(response.data);
-              observer.next(data);
-            });
-            // TODO Add onError and onComplete
-            break;
+    connection.then((socket: any) => {
```
Comment (dmitriyslaym) [inline packages/scalecube-microservice/src/ServiceCall/RemoteCall.ts]: socket is at least an object

**Pull request merged**